ENH: PeriodIndex now has period dtype

sinhrks opened the pull request:
- [x] split from #13755. 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

splitted `dtype` related part from #13755, to clarify the differences related to `PeriodIndex`

sinhrks committed:
ENH: PeriodIndex now has period dtype

MaximilianR commented:
Looks superb! Thanks @sinhrks

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -618,6 +618,37 @@ Furthermore:
+``PeriodIndex`` now has ``period`` dtype
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``PeriodIndex`` now has its own ``period`` dtype. ``period`` dtype is a
```
say pandas extension dtype

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -618,6 +618,37 @@ Furthermore:
+
+``PeriodIndex`` now has its own ``period`` dtype. ``period`` dtype is a
+original dtype which extends NumPy dtype like ``DatetimeTZDtype``
+for datetime with timezone. (:issue:`13941`)
```
``DatetimeTZDtype`` is really an internal thing, maybe just like: ``datetime64[ns, US/Eastern]`` or something

shoyer replied:
I might say "like categorical"

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,119 @@ def __eq__(self, other):
+        if freq is None:
+            raise ValueError("could not construct PeriodDtype")
+
+        from pandas._period import Period
```
this might be a perf issue; keep an eye here.

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,119 @@ def __eq__(self, other):
+        return "period[{freq}]".format(freq=self.freq.freqstr)
+
+    @property
+    def name(self):
```
side thing. We should just fix the metaclass to put in all of these inheritet methods....but that's another PR!

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,119 @@ def __eq__(self, other):
+    _metadata = ['freq']
+    _match = re.compile("period\[(?P<freq>.+)\]")
+
+    def __init__(self, freq, strict=False):
```
I think you have to do this in ``__new__`` (pretty much the same). As we need to return a singular object (and NOT a new instance everytime you do this).

```
In [7]: PeriodDtype.construct_from_string('period[D]')
Out[7]: period[D]

In [8]: p1 = PeriodDtype.construct_from_string('period[D]')

In [9]: p2 = PeriodDtype.construct_from_string('period[D]')

In [10]: p1 == p2
Out[10]: True

In [11]: p1 is p2
Out[11]: False

In [12]: from pandas.types.dtypes import DatetimeTZDtype

In [13]: t1 = DatetimeTZDtype.construct_from_string('datetime64[ns, US/Eastern]')

In [14]: t2 = DatetimeTZDtype.construct_from_string('datetime64[ns, US/Eastern]')

In [15]: t1 == t2
Out[15]: True

In [16]: t1 is t2
Out[16]: True
```

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,119 @@ def __eq__(self, other):
+    def __init__(self, freq, strict=False):
+        """
+        Parameters
+        ----------
```
I think accepting ``P|period[...]`` might be a good idea (as we name things ``Period``), but still display as ``period[D]`` is fine.

sinhrks replied:
As period without freq cannot determine its internal, you mean ``P[freq]`` or ``period[frqe]``? Or allow generic construction like ``np.datetime64`` without unit?

jreback replied:
no I just meant accepting Period[D] and period]D]

meaning case shouldn't matter for the P!

(so tiny change)

jreback replied:
still need freq of course

shoyer commented:
What do you get when you write `np.array(period_index)` or `period_index.values`? Currently these give back arrays of integers, which IMO is pretty broken.

jreback commented:
@shoyer yeah that should return an object array of Periods (like what a datetime tz does)

sinhrks commented:
split ``.values`` change as #13988, as it can be handled separately.

jreback commented:
@sinhrks a couple of minor changes, but otherwise lgtm. period is currently broken as its ``int64`` for its dtype. This makes things much more consistent (on the index side).

jorisvandenbossche commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -619,6 +619,36 @@ Furthermore:
+   In [1]: pi = pd.PeriodIndex(['2016-08-01'], freq='D')
+   In [2]: pi
+   Out[2]: PeriodIndex(['2016-08-01'], dtype='int64', freq='D')
+   In [3]: pd.types.api.is_integer_dtype(pi)
```
This is not the current public entry point (although depends on outcome of discussion in https://github.com/pydata/pandas/issues/13634), I think `pd.api.types.is_..` is

jreback replied:
yep.. @jorisvandenbossche right here

jorisvandenbossche commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -619,6 +619,36 @@ Furthermore:
+   pd.types.api.is_integer_dtype(pi)
+   pd.types.api.is_period_dtype(pi)
+
+   pi.dtype
```
Can you show this in the 'previous behaviour' ?

jorisvandenbossche commented on pandas/tseries/period.py:
```diff
@@ -147,6 +149,7 @@ class PeriodIndex(DatelikeOps, DatetimeIndexOpsMixin, Int64Index):
     second : int, array, or Series, default None
     tz : object, default None
         Timezone for converting datetime64 data to Periods
+    dtype : str or PeriodDtype, default None
```
dtype is already included above in the list of kwargs

sinhrks replied:
removed existing description based on args order.

jorisvandenbossche commented on pandas/tseries/period.py:
```diff
@@ -712,7 +735,8 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
             raise IncompatibleFrequency(msg)
-        return Index.get_indexer(self, target, method, limit, tolerance)
+        return Index.get_indexer(self, target, method,
+                                 limit, tolerance)
```
Does PEP8 implies to put it on one line?

jorisvandenbossche commented:
Cool!

Some first comments:

- I think this needs more documentation. Eg explaining that the freq is included in the dtype, that you can also do `dtype='period[freq]'`, more stress the API change in the dtype (it's no longer int). Further, also the narrative docs should be updated (not only the whatsnew).
- the repr is now a bit redundant (twice the 'D'):

  ```
In [25]: pi
Out[25]: PeriodIndex(['2016-08-01'], dtype='period[D]', freq='D')
  ```

  Leave out the freq? (although it is also useful to inform the user that the 'D' can be accessed as the freq ...)

- Is there a reason that `dtype_str` does not include the freq?

jreback commented:
could add some docs like this: http://pandas.pydata.org/pandas-docs/stable/timeseries.html#tz-aware-dtypes

sinhrks commented:
OK, added doc section to ``timeseries.rst``. Also fixed followings:

- ``dtype_str`` now uses the same repr as ``str(dtype)`` (use ``Index`` impl)
- ``astype`` related things

I don't have strong opinion about dtype repr, but it may better to keep it (different from timezone, users cannot distinguish freq if they don't know freq string).

jorisvandenbossche committed:
DOC: some clean-up of wording

jorisvandenbossche commented:
@sinhrks I had some comments on the wording, but thought it was maybe easier to just change it myself while reviewing than adding a bunch of minor comments -> https://github.com/sinhrks/pandas/pull/2. I hope that is OK

jorisvandenbossche commented:
The linter is failing on blank lines:

pandas/tests/types/test_dtypes.py:369:1: E303 too many blank lines (3)

jreback commented on doc/source/timeseries.rst:
```diff
@@ -1594,6 +1594,42 @@ objects:
+.. versionadded:: 0.19.0
+
+``PeriodIndex`` has a custom ``period`` dtype. This is a pandas extension
+dtype similar to the timezone aware dtype (``datetime64[ns, tz]``).
```
maybe put a reference in here as the datetimetz is below this in the docs.

jreback commented on doc/source/timeseries.rst:
```diff
@@ -1594,6 +1594,42 @@ objects:
+dtype similar to the timezone aware dtype (``datetime64[ns, tz]``).
+
+The ``period`` dtype holds the ``freq`` attribute and is represented with
+``period[freq]``, using :ref:`frequency strings <timeseries.offset_aliases>`.
```
add e.g. ``period[D]`` or ``period[M]``

jreback commented on doc/source/timeseries.rst:
```diff
@@ -1594,6 +1594,42 @@ objects:
+   pi
+   pi.dtype
+
+The ``period`` dtype can be used in ``.astype(...)``. It allows to change the
```
it allows one to change

jreback commented on doc/source/timeseries.rst:
```diff
@@ -1594,6 +1594,42 @@ objects:
+   dti = pd.date_range('2011-01-01', freq='M', periods=3)
+   dti
+   dti.astype('period[M]')
+
```
do we support the reverse (by ``.astype('datetime64[ns]')``)?

sinhrks replied:
Added.

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -623,6 +623,41 @@ Furthermore:
+
+``PeriodIndex`` now has its own ``period`` dtype. The ``period`` dtype is a
+pandas extension dtype which extends NumPy dtype like ``"category"`` (:issue:`13941`).
+As a consequence of this change, `PeriodIndex` no longer has an integer dtype:
```
double back-ticks around ``PeriodIndex``

jreback commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -623,6 +623,41 @@ Furthermore:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``PeriodIndex`` now has its own ``period`` dtype. The ``period`` dtype is a
+pandas extension dtype which extends NumPy dtype like ``"category"`` (:issue:`13941`).
```
these don't ``extend`` the ``NumPy`` dtype ``category`` (which doesn't exist), rather they are another pandas extension types like ``category``, ``datetime-tz``

jreback commented on pandas/core/algorithms.py:
```diff
@@ -424,12 +424,9 @@ def _value_counts_arraylike(values, dropna=True):
-                tz = orig.dt.tz
-            keys = DatetimeIndex._simple_new(keys, tz=tz)
-        if is_period:
+            from pandas.tseries.index import DatetimeIndex
```
I would move this import to where you do ``PeriodIndex`` for consistency

jreback commented on pandas/tests/types/test_cast.py:
```diff
@@ -241,12 +241,13 @@ def test_numpy_dtypes(self):
             # empty
             _find_common_type([])
 
-    def test_pandas_dtypes(self):
```
separate class for these?

sinhrks replied:
Added ``TestPandasDtype`` class.

jreback commented on pandas/tests/types/test_dtypes.py:
```diff
@@ -206,6 +208,162 @@ def test_parser(self):
+            dt = PeriodDtype(s)
+            self.assertEqual(dt.freq, pd.tseries.offsets.Day())
+            self.assertTrue(is_period_dtype(dt))
+
```
``self.assertTrue(dt is self.dtype)`` should be true

jreback replied:
nvm I see you have that below

jreback commented on pandas/tests/types/test_dtypes.py:
```diff
@@ -206,6 +208,162 @@ def test_parser(self):
+    def test_pandas_dtype(self):
+        self.assertEqual(pandas_dtype('period[D]'),
+                         PeriodDtype('period[D]'))
+        self.assertEqual(pandas_dtype('period[3M]'),
```
maybe some of these tests (in PeriodDtype) could be moved higher up to more generally test DatetimewithTZ ?

jreback commented on pandas/tests/types/test_dtypes.py:
```diff
@@ -206,6 +208,162 @@ def test_parser(self):
+        with tm.assertRaises(TypeError):
+            PeriodDtype.construct_from_string('foo')
+        with tm.assertRaises(TypeError):
+            PeriodDtype.construct_from_string('period[foo]')
```
add ``datetime64`` and ``datetime64[ns, US/Eastern]`` as additonal tests here (should raise)

sinhrks replied:
done.

jreback commented on pandas/tseries/period.py:
```diff
@@ -187,6 +190,16 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
             name = data.name
 
+        if dtype is not None:
+            dtype = pandas_dtype(dtype)
```
do we have sufficient pickle tests that exercise this construction?

sinhrks replied:
pickle already contains normal ``PeriodIndex``, is there any type specifically care?

- https://github.com/pydata/pandas/blob/master/pandas/io/tests/generate_legacy_storage_files.py#L75

jreback commented on pandas/tseries/period.py:
```diff
@@ -376,6 +389,11 @@ def _to_embed(self, keep_tz=False):
 
+    @property
+    def _int64index(self):
+        # do not cache, same as .asi8
```
why not cache?

sinhrks replied:
Followed the comment on ``numeric.py``. I can't find original issue for memory leak.

- https://github.com/pydata/pandas/blob/master/pandas/indexes/numeric.py#L135

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,123 @@ def __eq__(self, other):
+
+    @classmethod
+    def _parse_dtype_strict(cls, freq):
+        try:
```
you might want to:

```
try:
    if not freq.startswith('p'): 
        return None
    m = cls._match......
    if m is not None:
        .....
except:
    pass
return None
```

because otherwise ever time a ``is_period_dtype`` is called you end up raising (and catching) an exception which is pretty expensive. So better to do a quick test that will return for the common case w/o much work (common case is that it is NOT a period dtype if its a string).

shoyer replied:
Use `except AttributeError`. Never use a blanket `except`.

jreback commented:
just some minor changes / perf checks. Getting the ``is_period_type`` to be fast in the common case is important. For datetime w/tz too me several tries to make this happen

sinhrks committed:
doc / astype updates

chris-b1 commented on doc/source/timeseries.rst:
```diff
@@ -1594,6 +1594,47 @@ objects:
+.. _timeseries.timezone_series:
+
+The ``period`` dtype holds the ``freq`` attribute and is represented with
+``period[freq]`` like ``period[D]`` or ``period[D]``, using :ref:`frequency strings <timeseries.offset_aliases>`.
```
Second ``period[D]`` should be a different example?

jreback commented on pandas/types/dtypes.py:
```diff
@@ -244,6 +244,124 @@ def __eq__(self, other):
+    """
+    type = PeriodDtypeType
+    kind = 'O'
+    str = '|O08'
```
these don't mean anything just copied from numpy for compat

jreback commented:
thanks @sinhrks nice change!

sinhrks commented:
Thx for review. Because #14010 includes some period related tests, it'll broken on master. I'll send a fix soon.
